nvfp4 models stopped working on Apple M3

### What is the issue?

A recent update decided to wage a personal vendetta against my local Ollama setup with *-nvfp4 - Gemma4 in particular. The charming error message reads: "This library is using a deployment target (0x00020008) that is not supported on this OS." Very helpful, truly.

I'm not upset. I'm fine. The installer, in its infinite wisdom, chose the silent treatment - not a single word about breaking changes, just a casual drive by sabotage. No warning, no changelog pop-up, no "hey buddy, this might ruin your day." Just vibes and a broken setup.

### Relevant log output

```shell
% ollama run gemma4:26b-nvfp4
Error: 500 Internal Server Error: mlx runner failed: panic: mlx: Failed to load the default metallib. This library is using a deployment target (0x00020008) that is not supported on this OS. library not found library not found library not found  at /Users/runner/work/ollama/ollama/build/metal-v3/_deps/mlx-c-src/mlx/c/transforms.cpp:73 (exit: exit status 2)
```

### OS

macOS

### GPU

Apple

### CPU

Apple

### Ollama version

0.23.2

> [@johnkea](https://github.com/johnkea) can you post the logs? They should be in `~/.ollama/logs/server.log`.

time=2026-05-09T11:42:41.192+02:00 level=INFO source=model_recommendations.go:177 msg="model recommendations cache sleep scheduled" wait=4h9m18.299392913s consecutive_failures=0
[GIN] 2026/05/09 - 12:40:05 | 200 |      58.875µs |       127.0.0.1 | HEAD     "/"
[GIN] 2026/05/09 - 12:40:05 | 200 |   42.856208ms |       127.0.0.1 | POST     "/api/show"
[GIN] 2026/05/09 - 12:40:05 | 200 |    24.52825ms |       127.0.0.1 | POST     "/api/show"
time=2026-05-09T12:40:05.527+02:00 level=INFO source=sched.go:484 msg="system memory" total="36.0 GiB" free="25.4 GiB" free_swap="0 B"
time=2026-05-09T12:40:05.527+02:00 level=INFO source=sched.go:491 msg="gpu memory" id=0 library=Metal available="26.5 GiB" free="27.0 GiB" minimum="512.0 MiB" overhead="0 B"
time=2026-05-09T12:40:05.527+02:00 level=INFO source=client.go:359 msg="starting mlx runner subprocess" model=gemma4:26b-nvfp4 port=63712
time=2026-05-09T12:40:05.529+02:00 level=INFO source=sched.go:561 msg="loaded runners" count=1
time=2026-05-09T12:40:05.539+02:00 level=INFO source=server.go:44 msg="MLX engine initialized" "MLX version"=0.31.2-7-ge8ebdeb device=gpu
time=2026-05-09T12:40:05.609+02:00 level=INFO source=base.go:110 msg="Model architecture" arch=Gemma4ForConditionalGeneration
time=2026-05-09T12:40:06.183+02:00 level=INFO source=runner.go:159 msg="Loaded tensors from manifest" count=1498
panic: mlx: Failed to load the default metallib. This library is using a deployment target (0x00020008) that is not supported on this OS. library not found library not found library not found  at /Users/runner/work/ollama/ollama/build/metal-v3/_deps/mlx-c-src/mlx/c/transforms.cpp:73

goroutine 1 [running]:
github.com/ollama/ollama/x/internal/mlxthread.(*Thread).Do(0x1036c603b?, {0x10461e3a0?, 0x104869360?}, 0x1036b7b00?)
	/Users/runner/work/ollama/ollama/x/internal/mlxthread/thread.go:64 +0x60
github.com/ollama/ollama/x/mlxrunner.Execute({0x27197a0a8110, 0x4, 0x4})
	/Users/runner/work/ollama/ollama/x/mlxrunner/server.go:64 +0x428
github.com/ollama/ollama/runner.Execute({0x27197a0a80f0?, 0x0?, 0x0?})
	/Users/runner/work/ollama/ollama/runner/runner.go:22 +0xc4
github.com/ollama/ollama/cmd.NewCLI.func3(0x27197a58f000?, {0x1036b7908?, 0x4?, 0x1036b790c?})
	/Users/runner/work/ollama/ollama/cmd/cmd.go:2440 +0x50
github.com/spf13/cobra.(*Command).execute(0x27197a391b08, {0x27197a408820, 0x5, 0x5})
	/Users/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0x27197a732f08)
	/Users/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x2ec
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/Users/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:985
main.main()
	/Users/runner/work/ollama/ollama/main.go:12 +0x54
time=2026-05-09T12:40:06.190+02:00 level=ERROR source=sched.go:567 msg="error loading llama server" error="mlx runner failed: panic: mlx: Failed to load the default metallib. This library is using a deployment target (0x00020008) that is not supported on this OS. library not found library not found library not found  at /Users/runner/work/ollama/ollama/build/metal-v3/_deps/mlx-c-src/mlx/c/transforms.cpp:73 (exit: exit status 2)"
time=2026-05-09T12:40:06.190+02:00 level=INFO source=client.go:126 msg="stopping mlx runner subprocess" pid=91289
[GIN] 2026/05/09 - 12:40:06 | 500 |  703.887083ms |       127.0.0.1 | POST     "/api/generate"

--- a/scripts/build_darwin.sh
+++ b/scripts/build_darwin.sh
@@ -28,6 +28,24 @@ usage() {
 
 mkdir -p dist
 
+# Work around MLX's v3 metallib link leaking the macOS 26 deployment target.
+_relink_mlx_metallib() {
+    BUILD_DIR="$1"
+    KERNEL_DIR="$BUILD_DIR/_deps/mlx-build/mlx/backend/metal/kernels"
+    AIR_LIST="$BUILD_DIR/mlx-air-files.txt"
+    METALLIB="$KERNEL_DIR/mlx.metallib"
+
+    find "$KERNEL_DIR" -type f -name '*.air' | sort > "$AIR_LIST"
+    if [ ! -s "$AIR_LIST" ]; then
+        echo "error: could not find MLX AIR files in $KERNEL_DIR" >&2
+        exit 1
+    fi
+
+    status "Relinking MLX metallib"
+    rm -f "$METALLIB"
+    xargs xcrun -sdk macosx metallib -o "$METALLIB" < "$AIR_LIST"
+}
+
 
 ARCHS="arm64 amd64"
 while getopts "a:h" OPTION; do
@@ -83,6 +101,7 @@ _build_darwin() {
                 -DCMAKE_OSX_DEPLOYMENT_TARGET=14.0 \
                 -DCMAKE_INSTALL_PREFIX=$INSTALL_PREFIX
             cmake --build $BUILD_DIR --target mlx mlxc --parallel
+            _relink_mlx_metallib $BUILD_DIR
             cmake --install $BUILD_DIR --component MLX
 
             # Metal 4.x build (NAX-enabled, macOS 26+)

